cpp/src/arrow/compute/kernels/filter.cc

```diff
@@ -15,44 +15,468 @@
 // specific language governing permissions and limitations
 // under the License.
 
-#include "arrow/compute/kernels/filter.h"
+#include <algorithm>
+#include <memory>
+#include <utility>
+#include <vector>
 
-#include "arrow/array.h"
-#include "arrow/compute/kernel.h"
+#include "arrow/builder.h"
+#include "arrow/compute/context.h"
+#include "arrow/compute/kernels/filter.h"
+#include "arrow/util/bit-util.h"
+#include "arrow/util/checked_cast.h"
 #include "arrow/util/logging.h"
+#include "arrow/util/stl.h"
+#include "arrow/visitor_inline.h"
 
 namespace arrow {
-
 namespace compute {
 
-std::shared_ptr<DataType> FilterBinaryKernel::out_type() const {
-  return filter_function_->out_type();
+using internal::checked_cast;
+using internal::checked_pointer_cast;
+
+/// \brief BinaryKernel implementing Filter operation
+class ARROW_EXPORT FilterKernel : public BinaryKernel {
+ public:
+  explicit FilterKernel(const std::shared_ptr<DataType>& type) : type_(type) {}
+
+  Status Call(FunctionContext* ctx, const Datum& values, const Datum& filter,
+              Datum* out) override;
+
+  std::shared_ptr<DataType> out_type() const override { return type_; }
+
+  virtual Status Filter(FunctionContext* ctx, const Array& values,
+                        const BooleanArray& filter, int64_t length,
+                        std::shared_ptr<Array>* out) = 0;
+
+  static Status Make(const std::shared_ptr<DataType>& value_type,
+                     std::unique_ptr<FilterKernel>* out);
+
+ protected:
+  std::shared_ptr<DataType> type_;
+};
```

Comment:
I think you mean for this to be in filter.h? Can you add some docstrings?

Reply:
I can move it; I was going to follow `Cast()` and make only the [factory public](https://github.com/bkietz/arrow/blob/5a224b24ced09ed6692e0d0ef73a399dab072229/cpp/src/arrow/compute/kernels/cast.h#L62-L66) but I hadn't gotten to that yet